homeassistant/components/vacuum/roomba.py

```diff
@@ -0,0 +1,301 @@
+            )
+            self._vacuum.connect()
+
+        return self._vacuum
+
+    @asyncio.coroutine
+    def _try_command(self, mask_error, func, *args, **kwargs):
+        """Call a vacuum command handling error messages."""
+        # Error handling is not currently possible since Roomba()
+        # does not raise any exception on failure. But it will appear in the
+        # log.
+        # https://github.com/NickWaterton/Roomba980-Python/issues/8
+        yield from self.hass.async_add_job(partial(func, *args, **kwargs))
+        return True
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Turn the vacuum on."""
+        is_on = yield from self._try_command(
+            'Unable to start the vacuum: %s',
+            self.vacuum.send_command, 'start')
+        self._is_on = is_on
+
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs):
+        """Turn the vacuum off and return to home."""
+        yield from self.async_stop()
+        return_home = yield from self.async_return_to_base()
+        if return_home:
+            self._is_on = False
+
+    @asyncio.coroutine
+    def async_stop(self, **kwargs):
+        """Stop the vacuum cleaner."""
+        yield from self._try_command(
+            "Unable to stop: %s", self.vacuum.send_command, 'stop')
+
+    @asyncio.coroutine
+    def async_start_pause(self, **kwargs):
+        """Pause the cleaning task or replay it."""
+        if self.vacuum_state and self.is_on:
+            yield from self._try_command(
+                'Unable to start/pause (pause): %s',
+                self.vacuum.send_command, 'pause')
+        elif self._status == 'Stopped':
+            yield from self._try_command(
+                'Unable to start/pause (resume): %s',
+                self.vacuum.send_command, 'resume')
+        else:  # vacuum is off
+            yield from self._try_command(
+                'Unable to start/pause (start): %s',
+                self.vacuum.send_command, 'start')
+
+    @asyncio.coroutine
+    def async_return_to_base(self, **kwargs):
+        """Set the vacuum cleaner to return to the dock."""
+        return_home = yield from self._try_command(
+            'Unable to return home: %s', self.vacuum.send_command, 'dock')
+        if return_home:
+            self._is_on = False
+
+    @asyncio.coroutine
+    def async_send_command(self, command, params, **kwargs):
+        """Send raw command."""
+        _LOGGER.debug('async_send_command %s (%s), %s',
+                      command, params, kwargs)
+        yield from self._try_command(
+            "Unable to send command to the vacuum: %s",
+            self.vacuum.send_command, command, params)
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Fetch state from the device."""
+        # No data, no update
+        if not self.vacuum.master_state:
+            return
+        state = self.vacuum.master_state.get('state', {}).get('reported', {})
```

Comment:
I'm no async expert, but since the underlying library is using a thread to fetch update this should not be doing any IO. Am I right?